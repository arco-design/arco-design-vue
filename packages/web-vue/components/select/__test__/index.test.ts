import { mount } from '@vue/test-utils';
import Select from '../index';

describe('Select', () => {
  test('show dropdown correctly', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['Beijing', 'Shanghai', 'Guangzhou', 'Chengdu'],
      },
    });

    await wrapper.find('.arco-select-view').trigger('click');

    expect(document.body.outerHTML).toMatchSnapshot();
  });

  test('keyboard correctly', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['Beijing', 'Shanghai', 'Guangzhou'],
      },
    });

    const input = wrapper.find('input');
    await input.trigger('click');
    const dropdown = wrapper.findComponent({ name: 'SelectDropdown' });
    await input.trigger('keydown', { key: 'ArrowUp' });

    expect(dropdown.find('.arco-select-option-active').text()).toBe(
      'Guangzhou'
    );

    await input.trigger('keydown', { key: 'ArrowDown' });

    expect(dropdown.find('.arco-select-option-active').text()).toBe('Beijing');

    await input.trigger('keydown', { key: 'Enter' });

    expect(wrapper.emitted('change')?.[0]).toEqual(['Beijing']);
  });

  test('should support mouse', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['Beijing', 'Shanghai', 'Guangzhou'],
      },
    });

    const input = wrapper.find('input');
    await input.trigger('click');
    const dropdown = wrapper.findComponent({ name: 'SelectDropdown' });
    const options = dropdown.findAll('.arco-select-option');
    await options[0].trigger('mouseenter');
    await options[0].trigger('mouseleave');
    await options[1].trigger('mouseenter');
    await options[1].trigger('click');

    expect(wrapper.emitted('change')?.[0]).toEqual(['Shanghai']);
  });

  test('should show search option', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['Beijing', 'Shanghai', 'Guangzhou'],
        allowSearch: true,
      },
    });

    const input = wrapper.find('input');
    await input.trigger('click');
    await input.setValue('sh');
    const dropdown = wrapper.findComponent({ name: 'SelectDropdown' });
    const option = dropdown.find('.arco-select-option');
    expect(option.text()).toBe('Shanghai');
  });

  test('should show exact match search', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['AShanghai', 'BShanghai', 'Shanghai', 'Shan'],
        allowSearch: true,
      },
    });

    const input = wrapper.find('input');
    await input.trigger('click');
    await input.setValue('Shang');
    const dropdown = wrapper.findComponent({ name: 'SelectDropdown' });
    const option = dropdown.find('.arco-select-option');
    expect(option.text()).toBe('Shanghai');
  });

  test('should enable create option', async () => {
    const wrapper = mount(Select, {
      props: {
        options: ['Beijing', 'Shanghai', 'Guangzhou'],
        allowCreate: true,
      },
    });

    const input = wrapper.find('input');
    await input.trigger('click');
    await input.setValue('Xian');
    await input.trigger('keydown', { key: 'Enter' });

    expect(wrapper.emitted('change')?.[0]).toEqual(['Xian']);
  });
});
