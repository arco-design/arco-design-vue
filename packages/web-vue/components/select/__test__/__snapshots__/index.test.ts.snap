// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select show dropdown correctly 1`] = `"<body><div class=\\"arco-trigger-popup arco-trigger-position-bl\\" style=\\"z-index: 1001; pointer-events: auto; left: 0px; top: 4px; width: 0px;\\" trigger-placement=\\"bl\\"><transition-stub><div class=\\"arco-trigger-popup-wrapper\\" style=\\"transform-origin: 0 0;\\"><div class=\\"arco-trigger-content\\"><div class=\\"arco-select-dropdown\\"><!--v-if--><!--v-if--><div class=\\"arco-scrollbar arco-scrollbar-type-embed\\"><div class=\\"arco-scrollbar-container arco-select-dropdown-list-wrapper\\"><ul class=\\"arco-select-dropdown-list\\"><li class=\\"arco-select-option arco-select-option-active\\"><!--v-if--><span class=\\"arco-select-option-content\\">Beijing</span><!--v-if--></li><li class=\\"arco-select-option\\"><!--v-if--><span class=\\"arco-select-option-content\\">Shanghai</span><!--v-if--></li><li class=\\"arco-select-option\\"><!--v-if--><span class=\\"arco-select-option-content\\">Guangzhou</span><!--v-if--></li><li class=\\"arco-select-option\\"><!--v-if--><span class=\\"arco-select-option-content\\">Chengdu</span><!--v-if--></li></ul></div><!--v-if--><!--v-if--></div><!--v-if--></div></div><!----></div></transition-stub></div></body>"`;
